roles/confluent.variables/defaults/main.yml

```
@@ -291,28 +291,45 @@ zookeeper_custom_properties: "{{ zookeeper.properties }}"
 
 
 #### Kafka Broker Variables ####
-kafka_broker_default_listeners:
-  internal:
-    name: INTERNAL
-    port: 9092
-    ssl_enabled: "{{ ssl_enabled }}"
-    ssl_mutual_auth_enabled: "{{ ssl_mutual_auth_enabled }}"
-    sasl_protocol: "{{ 'OAUTH' if rbac_enabled|bool else sasl_protocol }}"
-
-kafka_broker_custom_listeners:
-  broker:
-    name: BROKER
-    port: 9091
-    ssl_enabled: "{{ ssl_enabled }}"
-    ssl_mutual_auth_enabled: "{{ ssl_mutual_auth_enabled }}"
-    sasl_protocol: "{{ sasl_protocol }}"
+
+kafka_broker_default_listeners: "{
+  'internal': {
+    'name': 'INTERNAL',
+    'port': 9092,
+    'ssl_enabled': {{ssl_enabled|string|lower}},
+    'ssl_mutual_auth_enabled': {{ssl_mutual_auth_enabled|string|lower}},
+    'sasl_protocol': '{% if rbac_enabled|bool %}OAUTH{% else %}{{sasl_protocol}}{% endif %}'
+  }{% if kafka_broker_configure_multiple_listeners|bool %},
+  'broker': {
+    'name': 'BROKER',
+    'port': 9091,
+    'ssl_enabled': {{ssl_enabled|string|lower}},
+    'ssl_mutual_auth_enabled': {{ssl_mutual_auth_enabled|string|lower}},
+    'sasl_protocol': '{{sasl_protocol}}'
+  }{% if kafka_broker_configure_control_plane_listener|bool %},
+  'controller': {
+    'name': 'CONTROLLER',
+    'port': 9089,
+    'ssl_enabled': {{ssl_enabled|string|lower}},
+    'ssl_mutual_auth_enabled': {{ssl_mutual_auth_enabled|string|lower}},
+    'sasl_protocol': '{{sasl_protocol}}'
+  }{% endif %}{% endif %}
+}"
+
+### Dictionary to put additional listeners to be configured within Kafka. Each listener must include a 'name' and 'port' key. Optionally they can include the keys 'ssl_enabled', 'ssl_mutual_auth_enabled', and 'sasl_protocol'
+kafka_broker_custom_listeners: {}
 
 # Deprecated variable
 kafka_broker_configure_additional_brokers: true
 
 ### Boolean to configure more than one kafka listener. Defaults to true. NOTE- kafka_broker_configure_additional_brokers is deprecated
 kafka_broker_configure_multiple_listeners: "{{kafka_broker_configure_additional_brokers}}"
 
+### Boolean to configure control plane listener on separate port, which defaults to 8089. Applied only if kafka_broker_configure_multiple_listeners is true
+kafka_broker_configure_control_plane_listener: false
+
+### Control Planer listener name.
+kafka_broker_control_plane_listener_name: "{{ 'controller' if kafka_broker_configure_control_plane_listener|bool and kafka_broker_configure_multiple_listeners|bool else kafka_broker_inter_broker_listener_name }}"
```

Ok this is looking good, final question, should this just be 
```suggestion
kafka_broker_control_plane_listener_name: controller
```

The if else *i think* is only necessary if you want to set the property always.

---

No. Initially this was set only to 'controller' but this break the build when `kafka_broker_configure_control_plane_listener` is `false` because the playbook fails with undefined variables. (`kafka_broker_listeners[kafka_broker_control_plane_listener_name]['name']` will be undefined in that case).

---

Ahh makes sense, that is one flaw with the combine dictionary solution that we have now... wish it didn't matter... let's call it tech debt!

I wonder if this would work:
```
control.plane.listener.name: "{{kafka_broker_listeners[kafka_broker_control_plane_listener_name]['name'] | default(kafka_broker_listeners[kafka_broker_inter_broker_listener_name]['name'])}}"
```

or
```
control.plane.listener.name: "{{kafka_broker_listeners[kafka_broker_control_plane_listener_name]|default(kafka_broker_listeners[kafka_broker_inter_broker_listener_name])['name']}}"
```

---

You're right that should work and it is possibly more elegant that using a conditional. Let's try to change it, will take only a few mins.

---

it might be 
```
control.plane.listener.name: "{{(kafka_broker_listeners[kafka_broker_control_plane_listener_name]['name'] | default(kafka_broker_listeners[kafka_broker_inter_broker_listener_name])).get('name')}}"
```

This might not be more elegant now that i try and type it out haha, wdyt?

---

I'm trying the first of the two suggestions. Just running molecule to check if it works, take a while on my laptop.

---

the first one is my vote (if it works)!

---

yes it does! :+1: